settings.lua

```diff
@@ -3763,6 +3702,325 @@ function OppositeTurnModeSetting:init(vehicle)
+	end
+end
+
+--use tool.move as if we are a player using mouse/axis ..
+function WorkingToolPositionsSetting.checkToolRotation(self,tool,toolIndex,posX,dt)
+	local spec = self.spec_cylindered
+	if tool.rotSpeed == nil then
+		return
+	end
+	
+	local curRot = { getRotation(tool.node) }
+	local newRot = curRot[tool.rotationAxis]
+	-- speed for frontloader, shovel, etc achses
+	local rotSpeed = 0.7
+	cpDiff = spec.cpWorkingToolPos[posX][toolIndex].curRot - newRot
+	if math.abs(cpDiff) > 0.03 then
+		if cpDiff < 0 then
+			rotSpeed=rotSpeed*(-1)
+		end
+	else 
+		tool.move = 0
+		return false
+	end
+	tool.move = rotSpeed
+	if tool.move ~= tool.moveToSend then
+		tool.moveToSend = tool.move
+		self:raiseDirtyFlags(spec.cylinderedInputDirtyFlag)
+	end
+    return true
+end
+
+--use tool.move as if we are a player using mouse/axis ..
+function WorkingToolPositionsSetting.checkToolTranslation(self,tool,toolIndex,posX,dt)	
+	local spec = self.spec_cylindered
+	if tool.transSpeed == nil then
+		return
+	end
+	
+	local curTrans = { getTranslation(tool.node) }
+	local newTrans = curTrans[tool.translationAxis]
+	-- speed for telescope arm
+	local transSpeed = 0.5
+	cpDiff = spec.cpWorkingToolPos[posX][toolIndex].curTrans - newTrans
+	if math.abs(cpDiff) > 0.05 then
+		if cpDiff < 0 then
+			transSpeed=transSpeed*(-1)
+		end
+	else 
+		tool.move = 0
+		return false
+	end
+	tool.move = transSpeed
+	if tool.move ~= tool.moveToSend then
+		tool.moveToSend = tool.move
+		self:raiseDirtyFlags(spec.cylinderedInputDirtyFlag)
+	end
+    return true
+end
+
+function WorkingToolPositionsSetting:isValidSpec(object)
+	if self.validSpecs == nil then 
+		return true
+	end	
+	for _,spec in pairs(self.validSpecs) do
+		if SpecializationUtil.hasSpecialization(spec, object.specializations)  then 
+			return true
+		end
+	end	
+end
+
+function WorkingToolPositionsSetting:hasValidToolPositions()
+	return #self.hasPosition == self.totalPositions
+end
+
+function WorkingToolPositionsSetting:getTotalPositions()
+	return self.totalPositions
+end
+
+function WorkingToolPositionsSetting:onWriteStream(streamId)
+	for i=1,self.totalPositions do 
+		streamWriteBool(streamId,self.hasPosition[i]==true)
+	end
+end
+
+function WorkingToolPositionsSetting:onReadStream(streamId)
+	for i=1,self.totalPositions do 
+		self.hasPosition[i] = streamReadBool(streamId)
+	end
+end
+
+function WorkingToolPositionsSetting:loadFromXml(xml, parentKey)
+	local value = getXMLBool(xml, self:getKey(parentKey))
+	if value then
+		for i=1,self.totalPositions do 
+			self.hasPosition[i] = true
+		end
+	end
+end
+
+function WorkingToolPositionsSetting:saveToXml(xml, parentKey)
+	setXMLBool(xml, self:getKey(parentKey), self:hasValidToolPositions())
+end
+
+function WorkingToolPositionsSetting:saveToXMLFile(xmlFile, key, usedModNames)
+	local spec = self.spec_cylindered
+	if spec.cpWorkingToolPos == nil or #spec.cpWorkingToolPos < spec.cpWorkingToolPosMax then 
```

Comment:
not sure about the history of this, is this for backward compatibility only or are we still storing CP related data in cpWorkingToolPos? What is it storing? And what is the WorkingToolPositionsSetting storing?

Reply:
Every object tool positions gets stored in the object directly as opposed to storing every object position in the main vehicle driver for easy saving and loading of these porsitions. "cpWorkingTool" is just a table I created for each object to store these positions.